### DISCUSSION TITLE
private key not found error even though I have loaded private key in strongswan.

### DISCUSSION THREAD

**Discussion by digambaringale333** [strongswan/strongswan#867] [Q&A]
1. I have a setup where I have installed CA and end point certificate on host.
2. Here is my connection config, I am using cert serial number as id in connection config.
 ```
[root@dipsechost1 different_ca]# swanctl --list-conn
host-host1: IKEv1/2, no reauthentication, rekeying every 14400s
  local:  192.168.43.254
  remote: 192.168.43.123
  local public key authentication:
    id: 95:7a:8b:7c:71:53:44:52
  remote public key authentication:
    id: %any
  net-net: TUNNEL, rekeying every 3600s
    local:  dynamic
    remote: dynamic

```
3. Here are my loaded certificates:
```
[root@dipsechost1 different_ca]# swanctl --list-certs | less

List of X.509 End Entity Certificates

  subject:  "C=I, ST=M, L=P, O=XYZ, OU=S, CN=CA_Signed_cluster01, E=support@xyz.com"
  issuer:   "C=I, O=XYZ, CN=RootCA1"
  validity:  not before Jan 28 20:08:12 2022, ok
             not after  Jan 28 20:08:12 2023, ok (expires in 364 days)
  serial:    95:7a:8b:7c:71:53:44:52
  flags:
  subjkeyId: 36:33:ec:28:bf:be:59:39:09:26:f4:3f:43:28:ef:b5:ab:4d:58:6c
  pubkey:    RSA 2048 bits, has private key
  keyid:     b3:b7:50:8d:67:f4:80:a3:3d:96:dd:9d:e0:75:a5:3f:20:19:06:b3
  subjkey:   36:33:ec:28:bf:be:59:39:09:26:f4:3f:43:28:ef:b5:ab:4d:58:6c

List of X.509 CA Certificates

  subject:  "C=I, O=XYZ, CN=RootCA1"
  issuer:   "C=I, O=XYZ, CN=RootCA1"
  validity:  not before Jan 28 19:34:18 2022, ok
             not after  Jan 26 19:34:18 2032, ok (expires in 3649 days)
  serial:    70:28:6d:03:41:e2:7a:f5
  flags:     CA CRLSign self-signed
  subjkeyId: ef:d0:58:37:20:af:c6:49:31:27:30:52:e8:de:40:a8:0f:0f:f3:57
  pubkey:    RSA 4096 bits
  keyid:     0d:48:33:5d:e0:b6:6d:6a:9f:0b:12:de:8e:fd:30:a4:96:f5:3c:25
  subjkey:   ef:d0:58:37:20:af:c6:49:31:27:30:52:e8:de:40:a8:0f:0f:f3:57


```
Above output shows that "has private key"

4. Tried to initiate the connection.
[root@dipsechost1 common_ca]# swanctl --initiate --ike host-host1
[IKE] initiating IKE_SA host-host1[3] to 192.168.43.123
[ENC] generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
[NET] sending packet: from 192.168.43.254[500] to 192.168.43.123[500] (392 bytes)
[IKE] retransmit 1 of request with message ID 0
[NET] sending packet: from 192.168.43.254[500] to 192.168.43.123[500] (392 bytes)
[NET] received packet: from 192.168.43.123[500] to 192.168.43.254[500] (293 bytes)
[ENC] parsed IKE_SA_INIT response 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) CERTREQ N(FRAG_SUP) N(HASH_ALG) N(CHDLESS_SUP) N(MULT_AUTH) ]
[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_AES128_XCBC/CURVE_25519
[IKE] received cert request for "C=I, O=XYZ, CN=RootCA1"
[IKE] sending cert request for "C=I, O=XYZ, CN=RootCA1"
[IKE] no private key found for '95:7a:8b:7c:71:53:44:52'
initiate failed: establishing IKE_SA 'host-host1' failed

not sure why it say's no private key found for '95:7a:8b:7c:71:53:44:52' (which is end point cert serial number).
Any help on this much appreciated.

**Comment by tobiasbrunner** [comment]
> I am using cert serial number as id in connection config

Why?

> not sure why it say's no private key found for '95:7a:8b:7c:71:53:44:52' (which is end point cert serial number).

Because that's what you configured as local identity, but it isn't an identity contained in the certificate. You want to use either the subject DN (which is the default if you configure a certificate but no identity), or a subjectAlternativeName contained in the certificate (yours doesn't contain any, though).

**Reply by digambaringale333** [reply]
I was referring to following page which talks about id parsing, there I can see that we can use serial numbers.
https://wiki.strongswan.org/projects/strongswan/wiki/IdentityParsing

Please correct me if I am looking at wrong info OR my understanding is incorrect.

**Reply by digambaringale333** [reply]
This was working few day's back but I don't know why it's not working now. Please have a look at following logs.

20:59:08 2022 05[IKE] <host-host1|1> received cert request for "C=I, O=XYZ, CN=Root CA"
20:59:08 2022 02[JOB] watcher got notification, rebuilding
20:59:08 2022 02[JOB] watcher going to poll() 7 fds
20:59:08 2022 02[JOB] watched FD 20 ready to write
20:59:08 2022 02[JOB] watcher going to poll() 6 fds
20:59:08 2022 05[IKE] <host-host1|1> sending cert request for "C=I, O=XYZ, CN=Root CA"
20:59:08 2022 02[JOB] watcher got notification, rebuilding
20:59:08 2022 02[JOB] watcher going to poll() 7 fds
20:59:08 2022 05[IKE] <host-host1|1> authentication of '4f:4a:4c:9c:d3:da:33:ad' (myself) with RSA_EMSA_PKCS1_SHA2_256 successful

**Reply by tobiasbrunner** [reply]
> I was referring to following page which talks about id parsing, there I can see that we can use serial numbers.

If you are referring to the table, then that's only for RDNs inside Distinguished Names (DNs). There is no IKEv2 identity type for serial numbers, nor can they be used as SANs.

However, what I wasn't aware of is that identities of type KEY_ID are actually matched against serial numbers of certificates. Those have to be configured as binary values, so make sure your config is correct.
